### PR TITLE
Batched energydensity integration 1

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

1.  Integration of EnergyDensityInput with EstimatorManagerInput.
2.  Handling of new EnergyDensityInput variant possibility by EstimatorManagerNew
3.  Trying to make the use of RealType and FullPrecRealType in the estimators by not using RealType = FullPrecRealType for classes that use FullPrecRealType for all floats.
4.  Add EnergyDensityInput to EstimatorManagerInputTesting. And update for its inclusion in test_EstimatorManagerNew.
5.  use std::bind consistently in MockGoldWalkerElements whether it really matters or not.
6.  Add integration test for EnergyDensityEstimator because correct reduction via EstimatorManagerNew is nontrivial.

Additional documentation for runtime use of EnergyDensityEstimator will be in the next PR.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Refactoring (no functional changes, no api changes)
- Build related changes
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation or build script changes

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

File: src/Estimators/tests/CMakeLists.txt [hunk 1]
```diff
@@ -80,11 +80,20 @@ endif()
...
-  #this is dependent on the directory creation and sym linking of earlier driver tests
-  set(SRCS FakeOperatorEstimator.cpp EstimatorManagerNewTest.cpp test_manager_mpi.cpp)
+  set(SRCS FakeOperatorEstimator.cpp
+      GenerateRandomParticleSets.cpp
+      MockGoldWalkerElements.cpp
+      EnergyDensityTest.cpp
+      EstimatorManagerNewTest.cpp
+      EstimatorManagerInputTest.cpp
+      test_EnergyDensityEstimatorIntegration.cpp
+      test_manager_mpi.cpp
```
Comment (ye-luo) [inline src/Estimators/tests/CMakeLists.txt:94]: I'm afraid there are too many classes being compiled twice. Could you introduce an intermediate target for the common cpp files.

**Review by ye-luo — COMMENTED** [review]
I found some confusing code that can be addressed in this PR or in a separate PR. I'd like to have the CMake issue preferably addressed in this PR.

File: src/Estimators/EnergyDensityEstimator.cpp [hunk 2]
```diff
@@ -77,8 +77,9 @@ NEEnergyDensityEstimator::NEEnergyDensityEstimator(const EnergyDensityInput& inp
     if (pset_static_)
     {
       auto [r_ptcls, z_ptcls] = extractIonPositionsAndCharge(*pset_static_);
-      spacegrids_.emplace_back(std::make_unique<NESpaceGrid<Real>>(spacegrid_inputs_[ig], ref_points_->get_points(), r_ptcls,
-                                                             z_ptcls, pset_dynamic_.getTotalNum(), N_EDVALS, periodic));
+      spacegrids_.emplace_back(std::make_unique<NESpaceGrid<Real>>(spacegrid_inputs_[ig], ref_points_->get_points(),
+                                                                   r_ptcls, z_ptcls, pset_dynamic_.getTotalNum(),
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/Estimators/EnergyDensityEstimator.cpp:81]: I will figure that out.  They relate to extra features from legacy that haven't been ported and won't be unless they are missed.

File: src/Estimators/EnergyDensityEstimator.cpp [hunk 3]
```diff
@@ -77,8 +77,9 @@ NEEnergyDensityEstimator::NEEnergyDensityEstimator(const EnergyDensityInput& inp
     if (pset_static_)
     {
       auto [r_ptcls, z_ptcls] = extractIonPositionsAndCharge(*pset_static_);
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/Estimators/EnergyDensityEstimator.cpp]: I've dropped this code this was used for the voronoi "grid" and I'm pursuing a different way of doing it that doesn't require a custom constructor.

File: src/Estimators/EnergyDensityEstimator.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/Estimators/EnergyDensityEstimator.cpp]: I've removed this unneeded constructor.

File: src/Estimators/tests/CMakeLists.txt (hunk 1 above)
Comment (PDoakORNL) [reply src/Estimators/tests/CMakeLists.txt]: I've added a shared target.

**Review by ye-luo — APPROVED** [review]

**Comment by prckent** [comment]
Test this please

**Pull request merged**